server/websocket.cpp

```diff
@@ -0,0 +1,213 @@
+
+net::io_context ioc{1};
+tcp::acceptor acceptor{ioc};
+uint16_t lastSocketID = 0;
+std::vector<std::shared_ptr<DfWebsocket>> sockets;
+
+static dfws::OnMessageHandler onMessageHandler = nullptr;
+static dfws::OnOpenHandler onOpenHandler = nullptr;
+static dfws::OnCloseHandler onCloseHandler = nullptr;
+
+void
+fail(beast::error_code ec, char const* what)
+{
+    std::cerr << what << ": " << ec.message() << "\n";
+}
+
+void dfws::SetOnMessage(OnMessageHandler h)
+{
+    onMessageHandler = h;
+}
+
+void dfws::SetOnOpen(OnOpenHandler h)
+{
+    onOpenHandler = h;
+}
+
+void dfws::SetOnClose(OnCloseHandler h)
+{
+    onCloseHandler = h;
+}
+
+class DfWebsocket : public std::enable_shared_from_this<DfWebsocket> {
+    websocket::stream<beast::tcp_stream> ws_;
+    beast::flat_buffer buffer_;
+
+public:
+    uint16_t socketID_;
+
+    // Take ownership of the socket
+    explicit
+    DfWebsocket(tcp::socket&& socket, uint16_t socketID)
+        : ws_(std::move(socket)), socketID_(socketID)
+    {
+        ws_.binary(true);
+    }
+
+    void start() {
+        // Set suggested timeout settings for the websocket
+        ws_.set_option(
+            websocket::stream_base::timeout::suggested(
+                beast::role_type::server));
+
+        // Set a decorator to change the Server of the handshake
+        ws_.set_option(websocket::stream_base::decorator(
+            [](websocket::response_type& res)
+            {
+                res.set(http::field::server,
+                    std::string(BOOST_BEAST_VERSION_STRING) +
+                        " websocket-server-async");
+            }));
+        // Accept the websocket handshake
+        ws_.async_accept(
+            beast::bind_front_handler(
+                &DfWebsocket::OnAccept,
+                shared_from_this()));
+    }
+
+    void DoRead() {
+        // Read a message into our buffer
+        ws_.async_read(
+            buffer_,
+            beast::bind_front_handler(
+                &DfWebsocket::OnRead,
+                shared_from_this()));
+    }
+
+    void OnAccept(beast::error_code ec)
+    {
+        if (ec)
+            return fail(ec, "accept");
+
+        onOpenHandler(socketID_);
+
+        DoRead();
+    }
+
+    void
+    OnRead(
+        beast::error_code ec,
+        std::size_t bytes_transferred)
+    {
+        boost::ignore_unused(bytes_transferred);
+
+        // This indicates that the session was closed
+        if (ec == websocket::error::closed)
+            return;
+
+        if (ec) {
+            if (ec.value() == boost::system::errc::operation_canceled) {
+                // this websocket has disconnected
+                // TODO delete it from the vector
+                onCloseHandler(socketID_);
+                return;
+            }
+            fail(ec, "read");
+        }
+
+        auto str = beast::buffers_to_string(buffer_.data());
+        onMessageHandler(socketID_, str);
+        buffer_.consume(buffer_.size());
+
+        DoRead();
```

Comment:
Same

Reply:
This call is synchronous, however the other one isn't so there is no infinite loop.